clang-tools-extra/clang-tidy/bugprone/IncorrectEnableSharedFromThisCheck.cpp

```diff
@@ -0,0 +1,62 @@
+//===--- IncorrectEnableSharedFromThisCheck.cpp - clang-tidy --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "IncorrectEnableSharedFromThisCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/DeclCXX.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+void IncorrectEnableSharedFromThisCheck::registerMatchers(MatchFinder *Finder) {
+  const auto EnableSharedFromThis =
+      cxxRecordDecl(hasName("enable_shared_from_this"), isInStdNamespace());
+  const auto QType = hasCanonicalType(hasDeclaration(
+      cxxRecordDecl(
+          anyOf(EnableSharedFromThis.bind("enable_rec"),
+                cxxRecordDecl(hasAnyBase(cxxBaseSpecifier(
+                    isPublic(), hasType(hasCanonicalType(
+                                    hasDeclaration(EnableSharedFromThis))))))))
+          .bind("base_rec")));
+  Finder->addMatcher(
+      cxxRecordDecl(
+          hasDirectBase(cxxBaseSpecifier(unless(isPublic()), hasType(QType))
+                            .bind("base")))
+          .bind("derived"),
+      this);
+}
+
+void IncorrectEnableSharedFromThisCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *BaseSpec = Result.Nodes.getNodeAs<CXXBaseSpecifier>("base");
+  const auto *Base = Result.Nodes.getNodeAs<CXXRecordDecl>("base_rec");
+  const auto *Derived = Result.Nodes.getNodeAs<CXXRecordDecl>("derived");
+  const bool IsEnableSharedFromThisDirectBase =
+      Result.Nodes.getNodeAs<CXXRecordDecl>("enable_rec") == Base;
+  const SourceRange ReplacementRange = BaseSpec->getSourceRange();
+  const std::string ReplacementString =
+      // BaseSpec->getType().getAsString() results in
+      // std::enable_shared_from_this<ClassName> or
+      // alias/typedefs of std::enable_shared_from_this<ClassName>
+      "public " + BaseSpec->getType().getAsString();
```

Comment:
The base type might be expanded from a macro, which we would not want to expand:

```c++
namespace std {
  template <typename T>
  struct enable_shared_from_this {};
}
struct A : std::enable_shared_from_this<A> {};
#define MACRO_A A
class B : MACRO_A {};

class B : public MACRO_A {}; // intended fix
class B : public A {}; // current fix
```

The solution would be to do a replacement on `SourceRange(Base->getBeginLoc(), Base->getBaseTypeLoc())` with only `public ` as the replacement string instead.
(Please add tests for this)

---

Note that this also has the added benefit of less potential for overlapping fixit ranges that could block each other.

Reply:
will implement soon when I have the bandwidth.

Reply:
approach confirmed to not work due to getBaseTypeLoc() not including the virtual or the access specifier. As a result, doing a blind SourceRange(BaseSpec->getBeginLoc(), BaseSpec->getBaseTypeLoc()) replacement with "public" just ends up with 

```
 //alias the template itself
 template <typename T> using esft_template = std::enable_shared_from_this<T>;
 
-class PrivateAliasTemplateClassBase : private esft_template<PrivateAliasTemplateClassBase> {};
+class PrivateAliasTemplateClassBase : public<PrivateAliasTemplateClassBase> {};
 //
 //
 
-class DefaultAliasTemplateClassBase : esft_template<DefaultAliasTemplateClassBase> {}; 
+class DefaultAliasTemplateClassBase : public<DefaultAliasTemplateClassBase> {}; 
 //
 //
 
 class PublicAliasTemplateClassBase : public esft_template<PublicAliasTemplateClassBase> {}; 
 
-struct PrivateAliasTemplateStructBase : private esft_template<PrivateAliasTemplateStructBase> {}; 
+struct PrivateAliasTemplateStructBase : public<PrivateAliasTemplateStructBase> {}; 
 //
 //

 struct A : std::enable_shared_from_this<A> {};
 #define MACRO_A A
-class B : MACRO_A {};
-class C : private MACRO_A {};
+class B : public {};
+class C : public {};
```

advice would be appreciated, because as it stands it doesn't seem possible to extract only the access specifier range. Current approach in my head is probably having a dedicated if statement just for macros, but that's kind of not ideal @5chmidti

Reply:
also, side note: MACRO_A doesn't count as direct inheritance either, so the fix won't emit on MACRO_A if we do the "issue FixItHint on direct inheritance only"

Reply:
> also, side note: MACRO_A doesn't count as direct inheritance either, so the fix won't emit on MACRO_A if we do the "issue FixItHint on direct inheritance only", which sort of resolves this comment indirectly as the issue is with FixItHint.

IMO macros are similar to aliases in this context, and we can do a replacement as well.

---

I found a way that works:

```diff
-  const SourceRange ReplacementRange = BaseSpec->getSourceRange();
-  const std::string ReplacementString =
-      // BaseSpec->getType().getAsString() results in
-      // std::enable_shared_from_this<ClassName> or
-      // alias/typedefs of std::enable_shared_from_this<ClassName>
-      "public " + BaseSpec->getType().getAsString();
+  const bool HasWrittenAccessSpecifier =
+      BaseSpec->getAccessSpecifierAsWritten() != AS_none;
+  const auto ReplacementRange = CharSourceRange(
+      SourceRange(BaseSpec->getBeginLoc(), BaseSpec->getBeginLoc()),
+      HasWrittenAccessSpecifier);
+  const llvm::StringRef Replacement =
+      HasWrittenAccessSpecifier ? "public" : "public ";
   const FixItHint Hint =
       IsEnableSharedFromThisDirectBase
-          ? FixItHint::CreateReplacement(ReplacementRange, ReplacementString)
+          ? FixItHint::CreateReplacement(ReplacementRange, Replacement)
```

with the following tests:
```c++
#define MACRO_PARAM(CLASS) std::enable_shared_from_this<CLASS>
class E : MACRO_PARAM(E) {};
// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: 'E' is not publicly inheriting from 'std::enable_shared_from_this', which will cause unintended behaviour when using 'shared_from_this'; make the inheritance public [bugprone-incorrect-enable-shared-from-this]
// CHECK-FIXES: class E : public MACRO_PARAM(E) {};
class F : private MACRO_PARAM(F) {};
// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: 'F' is not publicly inheriting from 'std::enable_shared_from_this', which will cause unintended behaviour when using 'shared_from_this'; make the inheritance public [bugprone-incorrect-enable-shared-from-this]
// CHECK-FIXES: class F : public MACRO_PARAM(F) {};
class G : public MACRO_PARAM(G) {};
```

Reply:
compiled and confirmed it works. Didn't know you can extract the SourceRange by feeding CharSourceRange getBeginLoc() twice (which is what surprised me the most, I would have thought it was getBeginLoc and getEndLoc) and the ITR boolean. Is this something that's learnable from the documentation otherwise?

Reply:
I only know of https://clang.llvm.org/docs/InternalsManual.html#sourcerange-and-charsourcerange, which is a rather short description. 
This is essentially saying: at the location (range), if there is a written access specifier, consider the range to be a token range (which will include the access specifier token in the range to be replaced), and if there is no access specifier written it is a character range (of size 0, so the replacement acts as an insertion).

---

nit: I forgot you can pass a single location to `SourceRange()`, and it will duplicate the location internally for begin and end.

Reply:
yes, will fix the nit and git merge origin to resolve the conflict.